syntax = "proto3";

// Define o pacote Go (opcional)
// option go_package = "./interfaces";

// Estrutura que representa um filme
message Movie {
    string id = 1;
    string plot = 2;
    repeated string genres = 3;
    int32 runtime = 4;
    repeated string cast = 5; 
    int32 num_mflix_comments = 6;
    string title = 7;
    string fullplot = 8;
    repeated string languages = 9;
    repeated string directors = 10;
    string rated = 11;
    string lastupdated = 12; // formato: 2015-08-13 00:46:30.660000000
    int64 year = 13;
    repeated string countries = 14;
    string type = 15;
}

// Estrutura para o nome de um filme
message MovieName {
    string nameMovie = 1;
}

// Estrutura para filtros de busca
message MovieFilters {
    repeated string values = 1;
}

// Estrutura para resposta com status e mensagens
message Response {
    int32 status = 1;
    string message = 2;
    Movie movie = 3;
    repeated Movie movies = 4;
}

// Estrutura para requisição com método e filtros
message Request {
    string method = 1;
    Movie movie = 2;
    MovieFilters filters = 3;
}

// Serviço que define os métodos para operações de CRUD e buscas
service MovieMethods {
    
    // Create
    rpc CreateMovie (Movie) returns (Response) {}

    // Retrieve (by ID or Name)
    rpc GetMovie (MovieName) returns (Response) {}

    // Retrieve by Actor
    rpc GetMoviesByActor (MovieFilters) returns (Response) {}

    // Retrieve by Category (Genre)
    rpc GetMoviesByCategory (MovieFilters) returns (Response) {}

    // Update
    rpc UpdateMovie (Movie) returns (Response) {}

    // Delete
    rpc DeleteMovie (MovieName) returns (Response) {}
}
